Documentation/design-docs/diagnostics/diagnostics-tools-contract.md

```diff
@@ -0,0 +1,26 @@
+# Diagnostics Tools Contract
+
+To perform its function, diagnostics tools need to make assumption on how the compiler, runtime and framework works. This was done in various obscure ways. Often, the dependency is taken in another code base such that breaking changes are detected only in runtime during integration test, which is often too late. 
+
+The crux of the problem is that the contract is implicit. Therefore, we are making changes to the product so that we establish a formal and explicit contract between the components and the diagnostics tools. For example, certain kind of breaking changes can now be detected during build time. You are reading this probably because you have made a change that broke the contract. In the sequel of the document, we will talk about why your change is breaking and what can you do about it.
+
+## Overall Principle
+At a breakpoint when the debuggee stops. We are freely look at any memory in the debuggee. This often the most convenient way to perform any debugging operation. Therefore, unlike other applications, debuggers are mostly interested in the fields of the debuggee, not the APIs. This is especially true in the AOT scenario where we do not have support for true function evaluation in the debuggee, yet.
+
+The debugger accesses the field through symbols, so this is mostly a no-op for the framework developers. But there are a few exceptions because the debugger is trying to make sense of those objects and give better diagnostics experience, for example:
+
+### Strings
+In a typical application, you have many strings. Strings are just another managed object such that we can inspect all the fields out of it. There is character array, we can inspect the characters. But that would be an awful experience. The user will want to just look at the string. In order to do that, the debugger make an assumption that System.String._firstChar represent the head of the character array and get the string out the debugee memory and display as a string. 
```

Comment:
You should enclose `System.String._firstChar` in "``" to get it displayed correctly. The underscore is otherwise interpreted as start of highlight - that's why there is several lines of italics.